### PR TITLE
[Dev Tools] Switch to Ava from tap

### PR DESCRIPTION
**Pull request by lsanwick**
I've switched the test runner to [ava](https://github.com/avajs/ava), which has similar ergonomics to `tap`,
but runs everything in parallel.  It also supports integration with [Wallaby](https://wallabyjs.com/) for a
great test writing environment in the editor.

Test Plan:

- [x] Unit tests pass!

### CODE REVIEW

File: wallaby.conf.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
+module.exports = function () {
```
Comment (Eomm) [inline wallaby.conf.js]: Why should we integrate this config?

I don't think it is worth including it

**Review by Eomm — COMMENTED** [review]
> but runs everything in parallel

What do you mean?
`tap` runs test in parallel by default: https://node-tap.org/docs/api/parallel-tests/ :

> The default jobs value for the command line runner is equal to the number of CPUs on your system, so it's as parallel as makes sense.

BUT:

> Within a single test file, the default jobs value is 1

So, maybe we should split the tests into multiple files for `ava` too?

---

Not against switching - I used `ava` before, and this module is not complex at all, but I want to know the real value here, or is it a matter of preference only? 🤣

**Comment by lsanwick** [comment]
All totally fair feedback! I haven't used `tap` in a very long time, so perhaps I made some more ignorant assumptions. Either way, test speed is not a problem for this repo :)

I use Wallaby for TDD development and it's honestly kinda life-changing with realtime feedback, code coverage, and time traveling debugging. Wallaby unfortunately doesn't support `tap`, but `ava` is supported and has a similar API.  The config file is required because `ava` isn't one of the test frameworks that is automatically detected and configured by Wallaby. 

Totally okay with skipping this PR if you feel strongly about it!

Just figured I did the work to debug and test the bugfix with this, so I might as well put it up for review :)

**Comment by Eomm** [comment]
I understand `ava` and it LGTM

What I don't agree with is the Wallaby - it is an opinionated tool. You can use it locally by adding it to the `gitignore`, but I don't want to push/advertise it through this repo

**Comment by lsanwick** [comment]
> I understand `ava` and it LGTM
> 
> What I don't agree with is the Wallaby - it is an opinionated tool. You can use it locally by adding it to the `gitignore`, but I don't want to push/advertise it through this repo

Done!

File: package.json [hunk 2]
```diff
@@ -8,8 +8,8 @@
     "lint": "standard && npm run lint:typescript",
     "lint:fix": "standard --fix",
     "lint:typescript": "standard --parser @typescript-eslint/parser --plugin @typescript-eslint/eslint-plugin types/**/*.ts",
-    "test": "tap test/**/*.test.js && tsd",
-    "test:unit": "tap test/**/*.test.js",
+    "test": "ava & tsd",
```
Comment (Eomm) [inline package.json]: It doesn't run the coverage by default.

I assume we need `c8 --100`

Comment (lsanwick) [reply package.json]: Good call, I've updated it.

File: package.json [hunk 3]
```diff
@@ -37,11 +37,12 @@
     "@types/node": "^20.1.0",
     "@typescript-eslint/eslint-plugin": "^5.46.0",
     "@typescript-eslint/parser": "^5.46.0",
+    "ava": "^5.3.1",
+    "c8": "8.0.0",
```
Comment (Eomm) [inline package.json]:
```diff
-    "c8": "8.0.0",
+    "c8": "^8.0.0",
```

**Review by Eomm — APPROVED** [review]

**Pull request merged**